### PR TITLE
Source selector to be current primary aware

### PR DESCRIPTION
**Pull request by mdarade**
Consensus messages would be interpreted by state transfer to
keep track of current primary replica so that it can be excluded
from list of sources.

### CODE REVIEW

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 1]
```diff
@@ -3040,6 +3020,22 @@ inline std::string BCStateTran::getSequenceNumber(uint16_t replicaId,
          std::to_string(chunkNum);
 }
 
+///////////////////////////////////////
+///// source selector comm'n
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp:3024]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 2]
```diff
@@ -3040,6 +3020,22 @@ inline std::string BCStateTran::getSequenceNumber(uint16_t replicaId,
...
+///////////////////////////////////////
+
+void BCStateTran::sniffConsensusMessage(MessageBase *msg) {
+  LOG_DEBUG(logger_, "Sniffed " << KVLOG(static_cast<MsgCode::Type>(msg->type()), msg->senderId()));
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp:3028]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 3]
```diff
@@ -3040,6 +3020,22 @@ inline std::string BCStateTran::getSequenceNumber(uint16_t replicaId,
...
+
+  if (msg->type() == MsgCode::PrePrepare) {
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp:3030]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 4]
```diff
@@ -3040,6 +3020,22 @@ inline std::string BCStateTran::getSequenceNumber(uint16_t replicaId,
...
+    sourceSelector_.updatePrimaryReplicaId(msg->senderId());
+  } else {
+    LOG_WARN(logger_, "Msg of type " << KVLOG(static_cast<MsgCode::Type>(msg->type())) << " not handled yet");
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp:3033]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/BCStateTran.hpp [hunk 5]
```diff
@@ -123,16 +125,37 @@ class BCStateTran : public IStateTransfer {
 
   void setEraseMetadataFlag() override { psd_->setEraseDataStoreFlag(); }
 
+  // TBD Make this configurable
+  // Consider msgs arrived from bft network only after interval duration of 2 seconds
+  void handoffConsensusMsg(MessageBase* msg) override {
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp:130]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/BCStateTran.hpp [hunk 6]
```diff
@@ -123,16 +125,37 @@ class BCStateTran : public IStateTransfer {
...
+    auto replica_id = msg->senderId();
+    auto itr = consensusMsgIntervalMap_.find(replica_id);
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp:132]: Thanks. Addressed.
Minimal filtering is there but code has been optimized now.

File: bftengine/src/bcstatetransfer/BCStateTran.hpp [hunk 7]
```diff
@@ -123,16 +125,37 @@ class BCStateTran : public IStateTransfer {
...
+    if (itr == consensusMsgIntervalMap_.end()) {
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp:133]: Thanks.

File: bftengine/src/bcstatetransfer/BCStateTran.hpp [hunk 8]
```diff
@@ -123,16 +125,37 @@ class BCStateTran : public IStateTransfer {
...
+      // discard frequent messages
+      delete msg;
+      return;
+    }
+    consensusMsgIntervalMap_.erase(itr);
+    handoff_->push(std::bind(&BCStateTran::sniffConsensusMessage, this, msg));
+  }
+
  protected:
+  std::unordered_map<uint16_t, Time> consensusMsgIntervalMap_;
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp:148]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/BCStateTran.hpp [hunk 9]
```diff
@@ -291,6 +309,7 @@ class BCStateTran : public IStateTransfer {
   ///////////////////////////////////////////////////////////////////////////
 
   SourceSelector sourceSelector_;
+  void sniffConsensusMessage(MessageBase*) override;
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp:312]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 10]
```diff
@@ -175,7 +177,39 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
   fetchingTimeStamp_ = 0;
   fetchRetransmissionCounter_ = 0;
   receivedValidBlockFromSrc_ = false;
+
+  LOG_INFO(logger_, "Selected new source replica " << currentReplica_);
+}
+
+void SourceSelector::updatePrimaryReplicaId(uint16_t newPrimary) {
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/SourceSelector.cpp:184]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 11]
```diff
@@ -175,7 +177,39 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
...
+  if (currentPrimary_ == newPrimary) return;
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/SourceSelector.cpp:185]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 12]
```diff
@@ -175,7 +177,39 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
...
+  auto clearCounters = [&]() { prePrepareMsgCntMap_.clear(); };
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/SourceSelector.cpp:186]: Thanks.

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 13]
```diff
@@ -175,7 +177,39 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
...
+
+  auto itr = prePrepareMsgCntMap_.find(newPrimary);
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/SourceSelector.cpp:188]: Code optimized.
Currently ST in unaware of view-change protocol and for now we would like to stay that way unless we get some major boost in performance.

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 14]
```diff
@@ -175,7 +177,39 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
...
+      clearCounters();
+    }
+    return;
+  }
+
+  if (++itr->second < prePrepareMsgMaxCnt_) return;
+  clearCounters();
+  LOG_TRACE(logger_, KVLOG(currentReplica_, currentPrimary_, newPrimary));
+
+  if (currentPrimary_ != NO_REPLICA and currentPrimary_ != newPrimary) {
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/SourceSelector.cpp:202]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 15]
```diff
@@ -175,7 +177,39 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
...
+    addPreferredReplica(currentPrimary_);
+  }
+  currentPrimary_ = newPrimary;  // Also handles case where currentPrimary == NO_REPLICA
+  removePreferredReplica(newPrimary);
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/SourceSelector.cpp:206]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 16]
```diff
@@ -175,7 +177,39 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
...
+
+  if ((currentReplica_ == newPrimary) or ((currentReplica_ == currentPrimary_ and currentPrimary_ == newPrimary))) {
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/SourceSelector.cpp:208]: Thanks. Addressed.

File: bftengine/src/bcstatetransfer/SourceSelector.hpp [hunk 17]
```diff
@@ -106,6 +120,8 @@ class SourceSelector {
   std::vector<uint16_t> actualSources_;
   bool receivedValidBlockFromSrc_;
 
+  std::unordered_map<uint16_t, uint64_t> prePrepareMsgCntMap_;
+  uint64_t prePrepareMsgMaxCnt_ = 15;  // considering we only sniff 1 msg every 2 sec
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/src/bcstatetransfer/SourceSelector.hpp:124]: Thanks. Addressed.

**Review by mdarade — COMMENTED** [review]
General comment:
1. All deletes are addressed through smart pointers. ST won't own any consensus msg. It would only extract interest info from consensus msg.
2. Counter based filtering is simple, optimized and it helps in situations where primary is changing quite frequently or in case of byzantine activity.

File: bftengine/src/bcstatetransfer/SourceSelector.hpp [hunk 18]
```diff
@@ -106,6 +121,9 @@ class SourceSelector {
   std::vector<uint16_t> actualSources_;
   bool receivedValidBlockFromSrc_;
 
+  uint16_t nominatedPrimary_ = NO_REPLICA;
+  uint16_t prePrepareMsgCounter_ = 0;
+  static constexpr uint16_t MAX_PREPREPARE_MSGS = 10;
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.hpp:126]: I think it is better to have this a configuration parameter.
Lets say an SRE finds some issue at costumer related to this value, and you want to increase it. It will be much easier for you to tell him to change the bft_config.yaml.
The work is very minimal to have it as a config parameter.
You can use for now internal default configuration, and not even propagate the parameter to bft_config.yaml (we can change it only once later on).

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 19]
```diff
@@ -175,7 +182,46 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
...
+    nominatedPrimary_ = NO_REPLICA;
+    prePrepareMsgCounter_ = 0;
+  };
+
+  if (nominatedPrimary_ == NO_REPLICA) {
+    nominatedPrimary_ = newPrimary;
+    return;
+  }
+
+  if (prePrepareMsgCounter_ < MAX_PREPREPARE_MSGS) {
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.cpp:204]: This block looks wrong in some cases, but I'm not running a unittest infront of me, so maybe i'm missing something :)

Lets see an example:
MAX_PREPREPARE_MSGS = 3
initial values nominatedPrimary_=NO_REPLICA, and currentPrimary_=0, prePrepareMsgCounter_=0
We have updateCurrentPrimary called multiple times with newPrimary=1,1,2 due to a view change and issues in the network (network partition):
1st call newPrimary with 1: prePrepareMsgCounter_=1, nominatedPrimary_=1 and exit in L201
2nd call with 1:  prePrepareMsgCounter_=2, exit on L206
3rd call with  2: prePrepareMsgCounter_=3, it jumps over block in L204 and didnt reset the counters! and chooses 2 as  currentPrimary_. This is wrong since 2 arrived only once.

If I was right, I suggest doing extensive checks in the source selector unittest, you have all the simple tools to easily write and test the code to handle all corner cases.

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 20]
```diff
@@ -175,7 +182,46 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
...
+    if (currentPrimary_ != NO_REPLICA) {
+      addPreferredReplica(currentPrimary_);
+    }
+    // Remove immediately as retransmission timeout, rotating source may end up choosing
+    // current primary as new source
+    removePreferredReplica(newPrimary);
+  }
+
+  currentPrimary_ = newPrimary;
+  LOG_INFO(logger_, KVLOG(currentPrimary_));
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.cpp:222]: I don't see the reason to have 2 logs in lines 222/210.
There is no difference in terms of performance and one will be always printed in any trace level, so why not printing all the extra info? it wouldn't hurt to have extra info always, it's a rare log...
Try to merge them both to be a single log.
`LOG_INFO(logger_, "New currentPrimary selected" ..... print old currentPrimary_, new currentPrimary_ (newPrimary), and currentReplica_`

File: bftengine/src/bcstatetransfer/SourceSelector.hpp [hunk 21]
```diff
@@ -54,6 +57,8 @@ class SourceSelector {
   // Return true if the source should be replaced, false otherwise.
   bool shouldReplaceSource(uint64_t currTimeMilli, bool badDataFromCurrentSource) const;
 
+  bool shouldReplaceSource() const { return (currentReplica_ == currentPrimary_); }
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.hpp:60]: Add a TODO - should not merge to prevent mistakes. This is temporary for UT I assume?

File: bftengine/src/bcstatetransfer/SourceSelector.hpp [hunk 22]
```diff
@@ -73,22 +78,32 @@ class SourceSelector {
 
   void addPreferredReplica(uint16_t replicaId) { preferredReplicas_.insert(replicaId); }
 
+  uint16_t currentPrimary() { return currentPrimary_; }
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.hpp:81]: We should write strange code for gtest.
Lets think of a solution to make the gest friend of source selector.
We want to get rid of one liners getters APIs in SS anyway.
i suggest to remove this one.
See the way we are using friendship with BCStateTran class from gtest.

File: bftengine/src/bcstatetransfer/SourceSelector.hpp [hunk 23]
```diff
@@ -73,6 +79,10 @@ class SourceSelector {
...
+
+  void removePreferredReplica(uint16_t replicaId) { preferredReplicas_.erase(replicaId); }
+
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/SourceSelector.hpp:84]: I dont see any place it is used else than the SS code.
see previous comment - we need to friend SS. That's the right solution.
We should try to write code while minimizing GTest impact.

File: bftengine/src/bcstatetransfer/SourceSelector.hpp [hunk 24]
```diff
@@ -106,6 +121,9 @@ class SourceSelector {
   std::vector<uint16_t> actualSources_;
   bool receivedValidBlockFromSrc_;
 
+  uint16_t nominatedPrimary_ = NO_REPLICA;
+  uint16_t prePrepareMsgCounter_ = 0;
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.hpp:125]: SS is anaware to the fact we use the report by fetching prepreapre. in that sense we want to keep it internal to BCStateTRan. I think a more general name is a better match:
`nominatedPrimaryCounter_`

in addition, I would put these 3 parameters straight after currentPrimary_, they all belong to the same logic , so it make sense ot put them one near another.

File: bftengine/src/bftengine/ReplicaForStateTransfer.hpp [hunk 25]
```diff
@@ -46,6 +48,14 @@ class ReplicaForStateTransfer : public IReplicaForStateTransfer, public ReplicaB
  protected:
   virtual void onTransferringCompleteImp(uint64_t checkpointNumberOfNewState) = 0;
 
+  template <typename T>
+  void consensusMessageHandler(MessageBase* msg) {
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaForStateTransfer.hpp:52]: Please move to cpp, keep only decleration.
That's why we have cpp. in that case hpp is used for declerations, cpp for defnitions.

File: bftengine/src/bftengine/ReplicaForStateTransfer.hpp [hunk 26]
```diff
@@ -46,6 +48,14 @@ class ReplicaForStateTransfer : public IReplicaForStateTransfer, public ReplicaB
...
+    if (msgs_to_sniff_.find(msg->type()) != msgs_to_sniff_.end()) {
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaForStateTransfer.hpp:53]: nit:
`msg_types_to_sniff_`  ??

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 27]
```diff
@@ -128,7 +128,12 @@ void ReplicaImp::registerMsgHandlers() {
 template <typename T>
 void ReplicaImp::messageHandler(MessageBase *msg) {
   T *trueTypeObj = new T(msg);
+  if (isCollectingState()) {
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp:131]: nit:
thinking quickly but didn't spent more than a minute on it: In general, it is not a good idea to handle a message straight to ST to sniff, you didn't validate the message?

File: bftengine/src/simplestatetransfer/SimpleStateTran.cpp [hunk 28]
```diff
@@ -89,6 +89,8 @@ class SimpleStateTran : public ISimpleInMemoryStateTransfer {
   void markUpdate(void* ptrToUpdatedRegion, uint32_t sizeOfUpdatedRegion) override;
   void setEraseMetadataFlag() override {}
 
+  virtual void handoffConsensusMessage(shared_ptr<ConsensusMsg>& msg) override{};
```
Comment (glevkovich) [inline bftengine/src/simplestatetransfer/SimpleStateTran.cpp:92]: See me prev. comment - should we have it as a pure interface or just define it in one of the lower classes?
Maybe we need ot discuss this.

**Review by glevkovich — CHANGES_REQUESTED** [review]

**Comment by glevkovich** [comment]
@mdarade, added some new comments, some of nit type (just a recommendation).
Reached till bcstatetransfer_tests.cpp, will continue tomorrow.

File: bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp [hunk 29]
```diff
@@ -630,6 +661,39 @@ TEST_F(BcStTest, dstFullStateTransfer) {
...
+    mockedSrc_->ReplyFetchBlocksMsg();
+    if (testParams_.expectedLastRequiredBlockNum <= config_.maxNumberOfChunksInBatch) break;
+    testParams_.expectedLastRequiredBlockNum -= config_.maxNumberOfChunksInBatch;
+    this_thread::sleep_for(chrono::milliseconds(20));
+    onTimerImp();
+
+    std::call_once(once_flag, [&] {
+      // For 7 intervals of itemData msgs we need to send 10 PrePrepare so that source gets changed.
+      auto msg = msg_generator_->GeneratePrePrepareMsg(ss.currentReplica());
+      for (uint16_t i = 1; i <= 10; i++) {
```
Comment (glevkovich) [inline bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp:683]: The number 7 can be ommitted, we might change the range collected. I assume you meant to say "for 7 batches"
Can write: "Generate prePrepare messages to trigger source selector to change the source to avoid primary "
Also:
I think it's not a good practice to put explicit numbers here.
You have MAX_PREPREPARE_MSGS (which might become config param , see other comment).
Extract it and use it - same comment for the rest of the tests written, some are using 9.

File: bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp [hunk 30]
```diff
@@ -630,6 +661,39 @@ TEST_F(BcStTest, dstFullStateTransfer) {
...
+        auto cmsg = make_shared<ConsensusMsg>(msg->type(), msg->senderId());
+        stateTransfer_->handoffConsensusMessage(cmsg);
+      }
+    });
+  }
```
Comment (glevkovich) [inline bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp:688]: I don't see any validation that source has been replaced? We only know that ST is done.

File: bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp [hunk 31]
```diff
@@ -630,6 +661,39 @@ TEST_F(BcStTest, dstFullStateTransfer) {
...
+  ASSERT_NO_FATAL_FAILURE(AssertFetchResPagesMsgSent());
+  bool doneSending = false;
+  while (!doneSending) mockedSrc_->ReplyResPagesMsg(doneSending);
+  // validate completion
+  ASSERT_TRUE(replica_.onTransferringCompleteCalled_);
+  ASSERT_EQ(BCStateTran::FetchingState::NotFetching, stateTransfer_->getFetchingState());
```
Comment (glevkovich) [inline bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp:694]: It would be good to call here getActualSources() and check that we indeed have 2 sources + after we have the metrics, check all are 0 and the one due to primary is 1.

File: bftengine/tests/bcstatetransfer/source_selector_test.cpp [hunk 32]
```diff
@@ -227,6 +227,48 @@ TEST_F(SourceSelectorTestFixture, select_amongst_preferred_replicas) {
   ASSERT_TRUE(source_selector.isPreferred(source_selector.currentReplica()));
 }
 
+TEST_F(SourceSelectorTestFixture, unknown_primary) {
+  auto source_selector =
+      SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs, maxFetchRetransmissions, GL);
+  source_selector.updateSource(kSampleCurrentTimeMs);  // would set up all preferred replicas internally.
+  ASSERT_TRUE(source_selector.isPreferred(source_selector.currentReplica()));
+  for (uint16_t i = 1; i <= 9; i++) {
```
Comment (glevkovich) [inline bftengine/tests/bcstatetransfer/source_selector_test.cpp:235]: Try to use the macro or config param MAX_PREPREPARE_MSGS.
If you don't have a way to extract it, please do not use hard coded numbers. put above a single constant MAX_PREPREPARE_MSGS with the same name and use it, its a better practice. as a last option you can also consider static assert to make sure they are equal.
this comment is valid for all new tests in file.

**Review by glevkovich — CHANGES_REQUESTED** [review]
Added here additional comments for tests.
I think we will have to discuss face to face some of the comments.

File: bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp [hunk 33]
```diff
@@ -373,7 +401,10 @@ class BcStTest : public ::testing::Test {
         itemDataMsg->requestMsgSeqNum = fetchBlocksMsg->msgSeqNum;
         itemDataMsg->dataSize = blk->totalBlockSize;
         memcpy(itemDataMsg->data, blk.get(), blk->totalBlockSize);
-        stateTransfer_->onMessage(itemDataMsg, itemDataMsg->size(), msg.to_, std::chrono::steady_clock::now());
+        stateTransfer_->onMessage(itemDataMsg,
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp:404]: Thanks. Addressed.

File: bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp (hunk 29 above)
Comment (mdarade) [reply bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp:683]: Thanks. Addressed.

File: bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp [hunk 34]
```diff
@@ -630,6 +661,39 @@ TEST_F(BcStTest, dstFullStateTransfer) {
...
+    testParams_.expectedLastRequiredBlockNum -= config_.maxNumberOfChunksInBatch;
+    this_thread::sleep_for(chrono::milliseconds(20));
+    onTimerImp();
+
+    std::call_once(once_flag, [&] {
+      // For 7 intervals of itemData msgs we need to send 10 PrePrepare so that source gets changed.
+      auto msg = msg_generator_->GeneratePrePrepareMsg(ss.currentReplica());
+      for (uint16_t i = 1; i <= 10; i++) {
+        auto cmsg = make_shared<ConsensusMsg>(msg->type(), msg->senderId());
+        stateTransfer_->handoffConsensusMessage(cmsg);
```
[the remark this answers is not in the document]
Comment (mdarade) [reply bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp:685]: Thanks. Addressed.

File: bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp (hunk 31 above)
Comment (mdarade) [reply bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp:694]: Thanks. Addressed.

File: bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp (hunk 30 above)
Comment (mdarade) [reply bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp:688]: Thanks. Addressed. DUP.

File: bftengine/tests/bcstatetransfer/source_selector_test.cpp (hunk 32 above)
Comment (mdarade) [reply bftengine/tests/bcstatetransfer/source_selector_test.cpp:235]: Thanks. Addressed.

File: bftengine/include/bftengine/IStateTransfer.hpp [hunk 35]
```diff
@@ -19,13 +19,29 @@
...
+class ConsensusMsg {
+ public:
+  ConsensusMsg(MsgType type, NodeIdType sender_id) : type_(type), sender_id_(sender_id) {}
+  MsgType type() { return type_; }
+  NodeIdType senderId() { return sender_id_; }
+
+ protected:
+  MsgType type_;
+  NodeIdType sender_id_;
+};
```
Comment (glevkovich) [inline bftengine/include/bftengine/IStateTransfer.hpp:43]: This one is simplified and shorter, and more correct in my opinion:
```
struct ConsensusMsg {
  ConsensusMsg() = delete;
  ConsensusMsg(MsgType type, NodeIdType sender_id) : type_(type), sender_id_(sender_id) {}
  const MsgType type_;
  const NodeIdType sender_id_;
};
```

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 36]
```diff
@@ -175,7 +187,49 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
   fetchingTimeStamp_ = 0;
   fetchRetransmissionCounter_ = 0;
   receivedValidBlockFromSrc_ = false;
+
+  LOG_INFO(logger_, "Selected new source replica " << currentReplica_);
+}
+
+void SourceSelector::updateCurrentPrimary(uint16_t newPrimary) {
+  if (currentPrimary_ == newPrimary) return;
+  ConcordAssert(isValidSourceId(newPrimary));
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.cpp:196]: Rethinking about this one,  might this be dangerous?
Should we assert? It is an external value, the source id is usually secured and validated by the communication module so at first it looks fine. But-  ST is unaware of that - who knows where it was fetched from in the future. If it is fetched from the message body it might be byzantine. Let's just replace it with an if + error + handle an error if happened.
If there was an error - I would reset all the counters for the nominated and exit - since it means something is very "unstable" and we want to replace primary when things are "stable".

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 37]
```diff
@@ -175,7 +187,49 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
...
+    nominatedPrimary_ = NO_REPLICA;
+    nominatedPrimaryCounter_ = 0;
+  };
+
+  if (nominatedPrimary_ == NO_REPLICA) {
+    nominatedPrimary_ = newPrimary;
+    return;
+  }
+
+  if ((newPrimary != nominatedPrimary_) and (nominatedPrimaryCounter_ < MAX_PREPREPARE_MSGS)) {
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.cpp:209]: I'm a bit confused here.
You still have MAX_PREPREPARE_MSGS but I saw you defined maxPrePrepareMessages as well.
Don't you think we should have it as a configuration parameter?
I lost my original comment so asking here again.

File: bftengine/src/bcstatetransfer/SourceSelector.cpp [hunk 38]
```diff
@@ -112,13 +118,15 @@ std::string SourceSelector::preferredReplicasToString() const {
   return oss.str();
 }
 
-bool SourceSelector::shouldReplaceSource(uint64_t currTimeMilli, bool badDataFromCurrentSource) const {
+SourceReplaceType SourceSelector::shouldReplaceSource(uint64_t currTimeMilli,
+                                                      bool badDataFromCurrentSource,
+                                                      bool lastInBatch) const {
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.cpp:123]: Adding a default just to save modifications in existing tests is not a good reason.
I don't see a use case here for a default (unless you needed to do huge modification test).
Production calls once and explicitly always.

**Review by glevkovich — CHANGES_REQUESTED** [review]
Few more comments

**Comment by glevkovich** [comment]
@mdarade , did you check why 6 checks are failing?

File: bftengine/src/bcstatetransfer/SourceSelector.cpp (hunk 38 above)
Comment (mdarade) [reply bftengine/src/bcstatetransfer/SourceSelector.cpp:123]: I believe Anand's PR would be addressing this as I've seen such comment there.

File: bftengine/src/bcstatetransfer/SourceSelector.hpp [hunk 39]
```diff
@@ -73,22 +85,34 @@ class SourceSelector {
...
 
   uint16_t currentReplica() const { return currentReplica_; }
 
   void onReceivedValidBlockFromSource();
 
   const std::vector<uint16_t> &getActualSources() { return actualSources_; }
 
+  void updateCurrentPrimary(uint16_t newPrimaryReplicaId);
+
+  uint16_t minPrePrepareMsgsForPrimaryAwarness() { return minPrePrepareMsgsForPrimaryAwarness_; }
```
Comment (glevkovich) [inline bftengine/src/bcstatetransfer/SourceSelector.hpp:106]: The impcat of the UT should be minimal on the production.
Here you create unneded getters, and potentially many more in the future - it looks akward if the reader is unaware of the reason. I think it should be removed,  the best idea is to add BcStTest as a friend of this class.

**Review by glevkovich — APPROVED** [review]
@mdarade , Thank you for the hard work.
Two last comments, but this one is approved.
Please do not merge it yet, lets wait for v1.5 freeze and you will probably also need to rebase on @ananpal changes + fix the small comments.

**Comment by mdarade** [comment]
Will be merging into dev branch.
